src/main/java/org/myrobotlab/service/Runtime.java

```diff
@@ -2006,11 +2006,19 @@ synchronized public static void unregister(String inName) {
 
     // FIXME - release autostarted peers ?
 
-    // last step - remove from registry
-    registry.remove(name);
+    // last step - remove from registry by making new registry
+    // thread safe way
+    Map<String, ServiceInterface> removedService = new TreeMap<>();

```

Comment:
A better way would be to use a `ConcurrentHashMap` and simply call `remove()` on it. If the key was already removed elsewhere, the call simply returns, no error thrown

Reply:
ConcurrentHashMap isn't sorted, and TreeMap isn't thread-safe, so I made it thread safe, the consequence is registry can't be final.